Mathlib/Analysis/Convex/EGauge.lean

```diff
@@ -171,24 +191,108 @@ lemma egauge_smul_right (h : c = 0 → s.Nonempty) (x : E) :
+    {U : ∀ i, Set (E i)} (hU : ∀ i ∈ I, Balanced 𝕜 (U i))
+    (x : ∀ i, E i) (hI₀ : I = univ ∨ (∃ i ∈ I, x i ≠ 0) ∨ (𝓝[≠] (0 : 𝕜)).NeBot) :
+    egauge 𝕜 (I.pi U) x = ⨆ i ∈ I, egauge 𝕜 (U i) (x i) := by
+  refine le_antisymm ?_ (iSup₂_le fun i hi ↦ le_egauge_pi hi _ _)
+  refine le_of_forall_lt' fun r hr ↦ ?_
+  have : ∀ i ∈ I, ∃ c : 𝕜, x i ∈ c • U i ∧ ‖c‖ₑ < r := fun i hi ↦
+    egauge_lt_iff.mp <| (le_iSup₂ i hi).trans_lt hr
+  choose! c hc hcr using this
+  obtain ⟨c₀, hc₀, hc₀I, hc₀r⟩ :
+      ∃ c₀ : 𝕜, (c₀ ≠ 0 ∨ I = univ) ∧ (∀ i ∈ I, ‖c i‖ ≤ ‖c₀‖) ∧ ‖c₀‖ₑ < r := by
+    have hr₀ : 0 < r := hr.bot_lt
+    have Hpos [hbot : (𝓝[≠] (0 : 𝕜)).NeBot] : ∃ c₀ ≠ (0 : 𝕜), ‖c₀‖ₑ < r :=
+      frequently_iff_neBot.mpr hbot |>.and_eventually
+        ((continuous_enorm.tendsto' 0 0 (by simp)).eventually_lt_const hr₀) |>.exists
+    rcases I.eq_empty_or_nonempty with rfl | hIne
+    · obtain hι | hbot : IsEmpty ι ∨ (𝓝[≠] (0 : 𝕜)).NeBot := by simpa [@eq_comm _ ∅] using hI₀
+      · use 0
+        simp [@eq_comm _ ∅, hι, hr₀]
+      · rcases Hpos with ⟨c₀, hc₀, hc₀r⟩
+        exact ⟨c₀, .inl hc₀, by simp, hc₀r⟩
+    · obtain ⟨i₀, hi₀I, hc_max⟩ : ∃ i₀ ∈ I, IsMaxOn (‖c ·‖ₑ) I i₀ :=
+        exists_max_image _ (‖c ·‖ₑ) hI hIne
+      by_cases H : c i₀ ≠ 0 ∨ I = univ
+      · exact ⟨c i₀, H, fun i hi ↦ by simpa [enorm] using hc_max hi, hcr _ hi₀I⟩
+      · push_neg at H
+        have hc0 (i : ι) (hi : i ∈ I) : c i = 0 := by simpa [H] using hc_max hi
+        have heg0 (i : ι) (hi : i ∈ I) : x i = 0 :=
+          zero_smul_set_subset (α := 𝕜) (U i) (hc0 i hi ▸ hc i hi)
+        have : (𝓝[≠] (0 : 𝕜)).NeBot := (hI₀.resolve_left H.2).resolve_left (by simpa)
+        rcases Hpos with ⟨c₁, hc₁, hc₁r⟩
+        refine ⟨c₁, .inl hc₁, fun i hi ↦ ?_, hc₁r⟩
+        simp [hc0 i hi]
+  refine egauge_lt_iff.2 ⟨c₀, ?_, hc₀r⟩
+  suffices x ∈ I.pi fun i ↦ c₀ • U i by
+    rcases hc₀ with hc₀ | rfl
+    · simpa only [← image_smul, ← piMap_image_pi (fun _ _ ↦ MulAction.surjective₀ hc₀)] using this
+    · simpa only [← image_smul, ← piMap_image_univ_pi] using this
+  intro i hi
+  exact (hU i hi).smul_mono (hc₀I i hi) (hc i hi)
+
+/-- The extended gauge of a point `x` in an indexed product with finite index type
+with respect to a product of balanced sets `U i`,
+is the supremum of the extended gauges of the components of `x`
+with respect to the corresponding balanced set.
+-/
+theorem egauge_univ_pi [Finite ι] {U : ∀ i, Set (E i)} (hU : ∀ i, Balanced 𝕜 (U i)) (x : ∀ i, E i) :
+    egauge 𝕜 (univ.pi U) x = ⨆ i, egauge 𝕜 (U i) (x i) :=
+  egauge_pi' finite_univ (fun i _ ↦ hU i) x (.inl rfl) |>.trans <| by simp
+
+/-- The extended gauge of a point `x` in an indexed product
+with respect to a product of finitely many balanced sets `U i`, `i ∈ I`,
+(and the whole spaces for the other indices)
+is the supremum of the extended gauges of the components of `x`
+with respect to the corresponding balanced set.
+
+This version assumes that `𝕜` is a nontrivially normed division ring.
```

Comment:
Am I right in thinking this is written as `[(𝓝[≠] (0 : 𝕜)).NeBot]`?

Reply:
@urkud, do you have any thoughts on this?

Reply:
I think that this is out of scope for this PR. Also, I think that the current situation with `ContinuousMul` (mixin) vs `NormedCommGroup` (extends) is a mess, but I don't know which approach is better for performance. As for this particular typeclass, I don't know if we should introduce `[NontriviallyNormed _]` mixin or `NontriviallyNormedRing`/`NontriviallyNormedDivisionRing`.